x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/custom/CustomService.java

```diff
@@ -0,0 +1,279 @@
+        ServiceUtils.validateInputTypeIsUnspecifiedOrInternal(inputType, validationException);
+    }
+
+    @Override
+    protected void doChunkedInfer(
+        Model model,
+        EmbeddingsInput inputs,
+        Map<String, Object> taskSettings,
+        InputType inputType,
+        TimeValue timeout,
+        ActionListener<List<ChunkedInference>> listener
+    ) {
+        listener.onFailure(new ElasticsearchStatusException("Chunking not supported by the {} service", RestStatus.BAD_REQUEST, NAME));
+    }
+
+    @Override
+    public Model updateModelWithEmbeddingDetails(Model model, int embeddingSize) {
+        if (model instanceof CustomModel customModel && customModel.getTaskType() == TaskType.TEXT_EMBEDDING) {
+            var newServiceSettings = getCustomServiceSettings(customModel, embeddingSize);
+
+            return new CustomModel(customModel, newServiceSettings);
+        } else {
+            throw new ElasticsearchStatusException(
+                Strings.format(
+                    "Can't update embedding details for model of type: [%s], task type: [%s]",
+                    model.getClass().getSimpleName(),
+                    model.getTaskType()
+                ),
+                RestStatus.BAD_REQUEST
+            );
+        }
+    }
+
+    private static CustomServiceSettings getCustomServiceSettings(CustomModel customModel, int embeddingSize) {
+        var serviceSettings = customModel.getServiceSettings();
+        var similarityFromModel = serviceSettings.similarity();
+        var similarityToUse = similarityFromModel == null ? SimilarityMeasure.DOT_PRODUCT : similarityFromModel;
+
+        return new CustomServiceSettings(
+            new CustomServiceSettings.TextEmbeddingSettings(
+                similarityToUse,
+                embeddingSize,
+                serviceSettings.getMaxInputTokens(),
+                serviceSettings.elementType()
+            ),
+            serviceSettings.getUrl(),
+            serviceSettings.getHeaders(),
+            serviceSettings.getQueryParameters(),
+            serviceSettings.getRequestContentString(),
+            serviceSettings.getResponseJsonParser(),
+            serviceSettings.rateLimitSettings(),
+            serviceSettings.getErrorParser()
+        );
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.ADD_INFERENCE_CUSTOM_MODEL;
+    }
+
+    public static class Configuration {
+        public static InferenceServiceConfiguration get() {
+            return configuration.getOrCompute();
+        }
+
+        private static final LazyInitializable<InferenceServiceConfiguration, RuntimeException> configuration = new LazyInitializable<>(
+            () -> {
+                var configurationMap = new HashMap<String, SettingsConfiguration>();
+                // TODO revisit this
```

Comment:
We'll need to create some more complex configuration types to support the fields (like maps, lists of lists etc). Maybe for now we don't expose this in the services API?